File: Algorithm/Greedy-Algorithm/JumpGame2.cpp
// leetcode 45

class Solution {
public:
    int jump(vector<int>& nums) {
        if(nums.size() < 2) {
            return 0;
        }
        
        int current_max_index = nums[0];
        int pre_max_index = nums[0];
        int jump_min = 1;
        
        for(int i = 0; i < nums.size(); i++) {
            if(current_max_index < i) {
                jump_min++;
                current_max_index = pre_max_index;
            }
            if(pre_max_index < nums[i] + i) {
                pre_max_index = nums[i] + i;
            }
        }
        return jump_min;
    }
};

File: Algorithm/Greedy-Algorithm/JumpGame2.java
package greedy;

/**
 * 
 * @description LeetCode 45, level: Hard
 *
 */
public class JumpGame2 {
    
    static int solution(int nums[]) {
        if(nums.length < 2) {
            return 0;
        }
        
        int jump = 1;
        // max step based on current position
        int currentMax = nums[0];
        // max step within currentMax
        int maxStep = nums[0];
        
        for(int i = 1; i < nums.length; i++) {
            int step = i + nums[i];
            if(i > currentMax) {
                jump++;
                currentMax = maxStep;
            }
            if(maxStep < step) {
                maxStep = step;
            }
        }
        
        return jump;
    }

    public static void main(String[] args) {
        int[] nums = {2, 3, 1, 1, 4};
        System.out.println(solution(nums));
    }

}
